src/inmanta/agent/handler.py

```diff
@@ -434,16 +435,74 @@ class ResourceHandler(object):
 
     def __init__(self, agent: "inmanta.agent.agent.AgentInstance", io: Optional["IOBase"] = None) -> None:
         self._agent = agent
-
         if io is None:
             raise Exception("Unsupported: no resource mgmt in RH")
         else:
             self._io = io
-
         self._client: Optional[protocol.SessionClient] = None
         # explicit ioloop reference, as we don't want the ioloop for the current thread, but the one for the agent
         self._ioloop = agent.process._io_loop
 
+    # Interface
+
+    def close(self) -> None:
+        pass
+
+    @abstractmethod
+    def deploy(
+        self,
+        ctx: HandlerContext,
+        resource: resources.Resource,
+        requires: Dict[ResourceIdStr, ResourceState],
+    ) -> None:
+        """
+        Main entrypoint of the handler that will be called by the agent to deploy a resource on the server.
+        This method is always called by the agent, even when one of the requires of the given resource
+        failed to deploy.
+
+        :param ctx: Context object to report changes and logs to the agent and server.
+        :param resource: The resource to deploy
+        :param requires: A dictionary mapping the resource id of each dependency of the given resource to its resource state.
+        """
+        pass
+
+    @abstractmethod
+    def execute(self, ctx: HandlerContext, resource: resources.Resource, dry_run: bool = False) -> None:
+        """
+        Update the given resource. This method is called by the agent. Most handlers will not override this method
+        and will only override :func:`~inmanta.agent.handler.ResourceHandler.check_resource`, optionally
```

Comment:
Also it references `ResourceHandler` methods rather than `HandlerAPI` ones. If you update this, please double check that they are all defined on the ABC.